### DISCUSSION TITLE
Can't persist entity with multiple entity manager Doctrine

### DISCUSSION THREAD

**Discussion by JoDrm** [symfony/symfony#49350] [Q&A]
I follow this manual to configure my multiple entity manager: https://symfony.com/doc/current/doctrine/multiple_entity_managers.html

My manager "customer" manage entity Client who is reference for client details.

My manager "workspace" manage entity User who is reference "uuid" of Client only.

So my doctrine.yml look like this : 
```
doctrine:
  dbal:
    default_connection: customer
    types:
      datetime_immutable: \Carbon\Doctrine\DateTimeImmutableType
      datetime: \Carbon\Doctrine\DateTimeType
    connections:
      workspace:
        # configure these for your database server
        driver: 'pdo_pgsql'
        server_version: '13'
        charset: utf8
        url: '%env(resolve:DATABASE_URL)%'
        wrapper_class: App\Doctrine\DynamicWorkspaceConnection
        logging: false
      customer:
        # configure these for your database server
        url: '%env(resolve:DATABASE_CAS_URL)%'
        driver: 'pdo_pgsql'
        server_version: '13'
        charset: utf8
  orm:
    default_entity_manager: customer
    entity_managers:
      workspace:
        connection: workspace
        naming_strategy: doctrine.orm.naming_strategy.underscore_number_aware
        mappings:
          Main:
            is_bundle: false
            type: attribute
            dir: '%kernel.project_dir%/src/Entity/Main'
            prefix: 'App\Entity\Main'
        filters:
          softdeleteable:
            class: Gedmo\SoftDeleteable\Filter\SoftDeleteableFilter
            enabled: true
          userteam_allowed:
            class: App\Filter\SQLFilter\AllowedByConnectedUserFilter
          user_allowed:
            class: App\Filter\SQLFilter\LicenceUserFilter
      customer:
        connection: customer
        naming_strategy: doctrine.orm.naming_strategy.underscore_number_aware
        mappings:
          Customer:
            is_bundle: false
            type: attribute
            dir: '%kernel.project_dir%/src/Entity/Cas'
            prefix: 'App\Entity\Cas'

```

My security.yml look like this :

```
security:
    # https://symfony.com/doc/current/security.html#c-hashing-passwords
  password_hashers:
    App\Entity\Cas\Client: 'auto'
  providers:
    clients:
      entity:
        class: App\Entity\Cas\Client
        manager_name: customer

  firewalls:
    dev:
      pattern: ^/(_(profiler|wdt)|css|images|js)/
      security: false

    auth:
      pattern: ^/
      stateless: true
      provider: clients
      user_checker: App\Security\Cas\ClientChecker
      entry_point: jwt
      json_login:
        check_path: login # The name in routes.yaml is enough for mapping
        username_path: email
        password_path: password
        success_handler: lexik_jwt_authentication.handler.authentication_success
        failure_handler: lexik_jwt_authentication.handler.authentication_failure
      jwt: ~
      refresh_jwt:
        check_path: api_refresh_token
```

**Signup Case :** 

When I signup in my app I use entity manager "customer" to persist Client object in database. It's work FINE.

**Login Cas :** 

When I login in my app it use entity manager "customer" to logged user with Client object. It's work FINE.

**Using "workspace" entity manager :** 

When I try to persist any object manage by "workspace" entity manager, I have this error

`Class 'App\\Entity\\Cas\\Client' was not found in namespaces configured in string App\\Entity\\Main.`

I think it's because when I authenticate the user on my API, the entity that's used to do that is Client, and it's handled only by my "customer" entity manager.

I don't know how to handle this problem, because if I cast $security->getUser() to type User, I can't persist with client entity manager.

Can I just switch the $security->getUser() dynamically when using "customer" or "workspace" manager ?

Do I use a chained provider ? Or custom provider ? Or use custom_authenticators fields ? 

Anyone have an idea to solve this problem?

**Comment by MatTheCat** [comment]
Did you read https://symfony.com/doc/current/doctrine/multiple_entity_managers.html? It says

> Entities cannot define associations across different entity managers.

Is it your case?

**Reply by JoDrm** [reply]
No I don't want to do that.

**Reply by MatTheCat** [reply]
We’ll need a [reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html) then.

**Reply by JoDrm** [reply]
DB manage by "workspace" contains User (id, uuid, roles, ... ).
DB manage by "customer" contains Client(id, uuid, firstname,lastname ....)  -> this DB is used to authenticated 

I copy manually `uuid` fields between, there have no DB relation manyToOne or ManyToMany ... ! 

But when I want to do a $workspaceEntityManager->persist($user) with "workspace" entity manager he want to manage Client entity. I don't understand why.

with this error : `Class 'App\\Entity\\Cas\\Client' was not found in namespaces configured in string App\\Entity\\Main.`

**Comment by ahundiak** [comment]
Please run `bin/console doctrine:mapping:info --em=customer` and post the relevant results.  Repeat for `--em=workspace` and maybe even `--em=default`.  Right now the suspicion is that you do have some of association between your entities or you somehow have the same entity type mapped in more than one manager.  And it is kind of hard to follow exactly what is being done by what.  The mapping info might clarify a few things.

**Reply by JoDrm** [reply]
bin/console doctrine:mapping:info --em=customer result 1 entity manage  :  
[OK]   App\Entity\Cas\Client

Client entity implement UserInterface

bin/console doctrine:mapping:info --em=workspace result 1 entity manage : 
 [OK]   App\Entity\Main\User\User

User entity doesn't implement UserInterface

"default" doesn't exist it's an error when write the post.

**Reply by ahundiak** [reply]
Thanks.  I only asked abut the default entity manager because you had referred to it a few times and wanted to make sure there was not a third manager lurking about.

Please confirm that `$security->getUser() ` returns an instance of Cas\Client and please clarify how you plan to use Main\User?  I don't understand the relationship.  You talked about `copying` stuff but I'm not sure if that was when mapping the entity or if you actually had two instances and are manually copying property values.  If you are dealing with instances then please provide a bit of code showing what you are doing.

**Reply by ghost** [reply]
Yes ` $security->getUser() ` return an instance of Cas\Client.

The logic of my app is Client can be create "Workspace" who a workspace is a dedicated DB with business logic, so in DB workspace we can find User(id, uuid ...), Role(code, name ...), Licence(code,name ...).
Yes in this DB logic User have relation with other entity. 
But any no with Client entity from other entity manager.

In DB CAS we retrieve Client(id, uuid, firstname, lastname ... ), and when I want to get user infos I use "customer" manager with $managerRegistry->getManager('customer')->getRepository(Client::class)->findBy($userUuid) where uuid is same in Cas DB and Workspace DB.

**Reply by ahundiak** [reply]
Well okay.  This is as far as I can go without seeing the entire code base.  Might help to start a new project and reproduce the error with as little code as possible and then post a link.

Entity managers are silos and there is no fundamental reason why flushing one entity manager will involve entities from the second one unless the entities reference each other.  The error you are getting is completely consistent with code like `$mainUser->setClient($casClient);` but you seems adamant that that is not happening.  Good luck.

And this is off-topic but it is possible to create classes like WorkspaceEntityManager and CustomerEntityManager (using decoration) and just typehint against them.  Saves you from having to use the registry and can help document your code.

**Reply by ghost** [reply]
Ok sorry for all that I just noticed the problem which comes from the Blameable attributes and that it tries to define the createdBy and updatedBy attributes with the Client object which is the one passed in $security->getUser().
Two days to notice it, Thank you for your answers and your time.

**Reply by ahundiak** [reply]
That is actually good to know.  Might want to add it as an answer to your stackoverflow question as well.
